Models/MuZero_agent_2.py

```diff
@@ -584,88 +589,136 @@ def player_to_play(player_num, action):
+            assert horizons > 0
+            reset_idx = (np.array(search_lens) % horizons == 0)
+            assert len(reset_idx) == num
+            is_reset_lst = reset_idx.astype(np.int32).tolist()
+            # reset_idx = arr of t/f
+            # reset_idx ==> is_reset_lst
+            # tree node.isreset = is_reset_list[node]
+            hidden_state_index_x += 1
+
+            # backpropagation along the search path to update the attributes
+            tree.batch_back_propagate(hidden_state_index_x, discount, value_prefix_pool, value_pool, policy_logits_pool,
+                                      min_max_stats_lst, results, is_reset_lst)
+    
+    def batch_policy(self, observation):
+        self.NUM_ALIVE = observation.shape[0] 
+        # Setup specialised roots datastruction, format: env_nums, action_space_size, num_simulations
+        roots_cpp = tree.Roots(self.NUM_ALIVE, config.ACTION_DIM, config.NUM_SIMULATIONS) 
+        network_output = self.network.initial_inference(observation)  
+
+        value_prefix_pool = np.array(network_output["value_logits"]).reshape(-1).tolist()
+        policy_logits_pool = np.array(network_output["policy_logits"]).tolist()
+
+        # prepare the nodes to feed them into batch_mcts
+        noises = [np.random.dirichlet([config.ROOT_DIRICHLET_ALPHA] * config.ACTION_DIM).astype(np.float32).tolist()
+                  for _ in range(self.NUM_ALIVE)]
+        roots_cpp.prepare(config.ROOT_EXPLORATION_FRACTION, noises, value_prefix_pool, policy_logits_pool)
         
-        action = [int(self.select_action(root[i])) for i in range(self.NUM_ALIVE)]
-        
-        # Masking only if training is based on the actions taken in the environment.
-        # Training in MuZero is mostly based on the predicted actions rather than the real ones.
-        # network_output["policy_logits"], action = self.maskInput(network_output["policy_logits"], action)
+        # Output for root node
+        hidden_state_pool = network_output["hidden_state"] 
+
+        self.run_batch_mcts(roots_cpp, hidden_state_pool)  # set up nodes to be able to find and select actions 
+        roots_distributions = roots_cpp.get_distributions()
+        actions = [] 
+        # This variable controls if distributions is randomly created, such as during the very first loop, however it doesnt look like it always being True impacts anything 
+        start_training = True 
+        temp = self.visit_softmax_temperature() # controls the way actions are chosen
+        for i in range(self.NUM_ALIVE):
+            deterministic = False # False = sample distribution, True = argmax 
+            if start_training:
+                distributions = roots_distributions[i]
+            else:
+                #random distributions if training has just started 
+                distributions = np.ones(config.ACTION_DIM)
+            action, entropy = self.select_action(distributions,temperature=temp,deterministic=deterministic) 
```

Comment:
lets not return entropy if we are never going to use it.

Reply:
I think you were suggesting that entropy be my next ticket to work on, so we will be using it shortly